core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/AST/AndPropertyShape.java

```diff
@@ -227,15 +227,45 @@ public PlanNode getAllTargetsPlan(ConnectionsGroup connectionsGroup, boolean neg
 
 	@Override
 	public String buildSparqlValidNodes(String targetVar) {
-		return and.stream()
-				.map(propertyShapes -> propertyShapes
-						.stream()
-						.map(propertyShape -> propertyShape.buildSparqlValidNodes(targetVar))
-						.reduce((a, b) -> a + "\n" + b))
-				.filter(Optional::isPresent)
-				.map(Optional::get)
-				.reduce((a, b) -> a + "\n" + b)
-				.orElse("");
+
+		if (hasOwnPath()) {
+			// within property shape
+			String objectVariable = "?b";
+			String pathQuery1 = getPath().getQuery(targetVar, objectVariable, null);
+
+			String collect = and.stream()
+					.map(l -> l.stream()
+							.map(p -> p.buildSparqlValidNodes(objectVariable))
+							.reduce((a, b) -> a + " && " + b))
+					.filter(Optional::isPresent)
+					.map(Optional::get)
+					.collect(Collectors.joining(" ) && ( ", "( ",
+							" )"));
+
+			String query = pathQuery1 + "\n FILTER (! EXISTS {\n" + pathQuery1.replaceAll("(?m)^", "\t")
+					+ "\n\tFILTER(!(" + collect + "))\n})";
+
+			String pathQuery2 = getPath().getQuery(targetVar, randomVariable(), null);
+
+			query = "{\n#VALUES_INJECTION_POINT#\n " + query.replaceAll("(?m)^", "\t")
```

Comment:
what is `VALUES_INJECTION_POINT`?

Reply:
It's where we inject SPARQL values. Since we have unions we end up introducing a new scope, which leads to poor performance. So we add points where we can inject sparql values, for the cases where we need to run the query against a subset of nodes.

Reply:
Ah I see. ObNitpick: it might be good to declare a constant for this substitution pattern.